### PR TITLE
fix Issue 22929 - importC: extern array with unknown length gives bou…

### PR DESCRIPTION
**Pull request by WalterBright**
…nds errors

This was actually two problems. The first test case is handled by lines optimize.c lines 555-561, the second test case the rest.

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -8472,6 +8472,19 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                     // OR it in, because it might already be set for C array indexing
                     exp.indexIsInBounds |= bounds.contains(getIntRange(exp.e2));
                 }
+                else if (auto ve = exp.e1.isVarExp())
+                {
+                    if (sc.flags & SCOPE.Cfile && t1b.ty == Tsarray)
```
Comment (RazvanN7) [inline src/dmd/expressionsem.d:8477]: Micro-optimization: swap this condition with the one above. Compiling C files occurs less frequent than compiling D files.

Also, I am a bit worried for the effect that all of these C file conditions have on the compilation of normal D code.

File: src/dmd/optimize.d [hunk 2]
```diff
@@ -504,6 +504,17 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
                     {
                         eint = ei;
                     }
+                    else if (auto se = ep.e1.isSymOffExp())
+                    {
+                        if (!se.var.isReference() &&
+                            !se.var.isImportedSymbol() &&
+                            se.var.isDataseg() &&
+                            se.var.isCsymbol())
```
Comment (RazvanN7) [inline src/dmd/optimize.d:512]: In this situation I have to mentally load myself with all of the previous conditions and then at the end I see that it applies only to C symbols. Why not start with this condition(preferably put it in the outer if)? That way, it becomes clear that what follows applies to C land. This should be a general rule.

File: src/dmd/expressionsem.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/expressionsem.d:8477]: ok.

I work to minimize these changes.

File: src/dmd/optimize.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/optimize.d]: I realized the isCsymbol() isn't necessary and removed it.

**Comment by WalterBright** [comment]
This is blocking fixing https://issues.dlang.org/show_bug.cgi?id=22935 which is strongly related.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**